documentation/adr/0003-use-mqtt-to-publish-trip-modifications.md

```diff
@@ -0,0 +1,62 @@
+# 3. Use MQTT to publish Trip Modifications
+
+Date: 2024-07-04
+
+## Status
+
+Proposed
+
+## Context
+
+We need to send [GTFS Trip Modifications](https://gtfs.org/realtime/feed-entities/trip-modifications/)
+to Transit Data.
+
+## Decision
+
+To minimize the amount of resources used and latency between Skate publishing or
+un-publishing a Trip Modification and Transit Data becoming aware, we are going
+to use MQTT to form an event stream between Skate and Transit Data.
+
+Since this is an output of Skate, we will not make this service required to do
+local development.
+
+### Alternatives Considered
+
+#### Polling based alternatives
+1. HTTP API
+  - We _could_ host an internal API on Skate which Transit Data can Poll. A
+  benefit of this is we can "render straight from the database" and avoid any
+  sort of expiration/synchronization logic with an external service.
+  Additionally, it is already an established practice within Transit Data to
+  poll an API (Ex: `Concentrate` polls `RTR`).  
+  The downside being that we are then requiring extra resources to continuously
+  poll, at 1s intervals, to check for any changes. Additionally, Skate's current
+  access control mechanism is concerned entirely with _users_ and not providing
+  pragmatic access. Using an external service allows us to push authentication
+  for this access to other services and IaC instead of requiring that Skate also
+  concerns itself with these issues.
```

Comment:
Another worthwhile point is that an API would be a good choice if "reads" were infrequent while "writes" were frequent, but the reality is the other way around - "reads" happen once a second, or any time a user wants to look something up, while write will only happen a few times a week

Some possible wordsmithing (no need to take this suggestion exactly):

```suggestion
  - We considered delivering this information by way of an API that
    Transit Data could poll. That would have the benefit of allowing
    us to "render straight from the database" and avoid any sort of
    expiration/synchronization logic with an external service.
    Additionally, it is already an established practice within Transit
    Data to poll an API (Ex: `Concentrate` polls `RTR`).

    This might be an appropriate choice if we expected there to be
    more "writes" than "reads" -- that is, if we expected officials to
    be modifying detours more frequently than Transit Data would be
    querying them. But we actually expect many more "reads" than
    "writes" - we'll need to access this data any time a rider looks
    up a bus route, or any time an operator loads turn-by-turn
    directions, or possibly once per second, while writes will only
    happen a few times a week, when an official is creating or
    modifying a detour. Creating an API would create performance
    concerns that messaging wouldn't.

    Additionally, Skate's current access control mechanism is
    concerned entirely with _users_ and not providing pragmatic
    access. Using an external service allows us to push authentication
    for this access to other services and IaC instead of requiring
    that Skate also concerns itself with these issues.
```

Rendered:
>  - We considered delivering this information by way of an API that
    Transit Data could poll. That would have the benefit of allowing
    us to "render straight from the database" and avoid any sort of
    expiration/synchronization logic with an external service.
    Additionally, it is already an established practice within Transit
    Data to poll an API (Ex: `Concentrate` polls `RTR`).
>
>    This might be an appropriate choice if we expected there to be
    more "writes" than "reads" -- that is, if we expected officials to
    be modifying detours more frequently than Transit Data would be
    querying them. But we actually expect many more "reads" than
    "writes" - we'll need to access this data any time a rider looks
    up a bus route, or any time an operator loads turn-by-turn
    directions, or possibly once per second, while writes will only
    happen a few times a week, when an official is creating or
    modifying a detour. Creating an API would create performance
    concerns that messaging wouldn't.
>
>    Additionally, Skate's current access control mechanism is
    concerned entirely with _users_ and not providing pragmatic
    access. Using an external service allows us to push authentication
    for this access to other services and IaC instead of requiring
    that Skate also concerns itself with these issues.